### PR TITLE
Feature/redpanda

### PR DESCRIPTION
**Pull request by senior7515**

### CODE REVIEW

**Comment by noahdesu** [comment]
nice. i think in the short term it'd be nice to see two modes: (a) build all dependencies and (b) build the minimum number of deps based on lowest common denominator distro which might be something like fmt, dpdk, seastar.

**Comment by senior7515** [comment]
```
➜  build git:(master) ✗ cmake ..                                                                                                                                                                          [20/9790]
-- Building with ccache: /bin/ccache, CCACHE_DIR=/dev/shm/ccache
-- RP_ENABLE_INTEGRATION_TESTS=ON                                                          
-- RP_ENABLE_UNIT_TESTS=ON                                                                                                   
-- RP_ENABLE_BENCHMARK_TESTS=ON                                                                            
-- RP_ENABLE_HONEY_BADGER_TESTS=ON
-- Configuring done
-- Generating done
-- Build files have been written to: /home/agallego/workspace/v/build/v_deps_build
[  3%] Built target xy                                                                                       
[  6%] Built target zlib                                                                                    
[  9%] Built target xml2                                                                         
[ 12%] Built target libpciaccess                                        
[ 14%] Built target GSL              
[ 17%] Built target numactl                
[ 20%] Built target zstd           
[ 23%] Built target gflags
[ 26%] Built target xxhash
[ 29%] Built target hwloc                  
[ 32%] Built target fmt                                                
[ 35%] Built target boost                                                                                                    
[ 37%] Built target dpdk
[ 40%] Built target protobuf
[ 43%] Built target lksctp-tools
[ 46%] Built target yaml-cpp
[ 49%] Built target lz4
[ 52%] Built target c-ares
[ 55%] Built target cryptopp
[ 58%] Built target gmp
[ 60%] Built target nettle
[ 63%] Built target gnutls
[ 66%] Built target seastar
[ 68%] Built target googlebenchmark
[ 71%] Built target HdrHistogram
[ 74%] Built target gtest
[ 77%] Built target glog
[ 80%] Built target flatbuffers
[ 80%] Performing configure step for 'smf'
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   filesystem
--   program_options
--   thread
--   unit_test_framework
--   system
--   chrono
--   date_time
--   atomic
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find c-ares: Found unsuitable version "", but required is at
  least "1.13" (found
  /home/agallego/workspace/v/build/v_deps_install/lib/libcares.so)
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:376 (_FPHSA_FAILURE_MESSAGE)
  /home/agallego/workspace/v/build/v_deps_install/lib64/cmake/Seastar/Findc-ares.cmake:47 (find_package_handle_standard_args)
  /home/agallego/workspace/v/build/v_deps_install/lib64/cmake/Seastar/SeastarConfig.cmake:41 (find_package)
  CMakeLists.txt:51 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/agallego/workspace/v/build/v_deps_build/smf-prefix/src/smf-build/CMakeFiles/CMakeOutput.log".
See also "/home/agallego/workspace/v/build/v_deps_build/smf-prefix/src/smf-build/CMakeFiles/CMakeError.log".
gmake[2]: *** [CMakeFiles/smf.dir/build.make:118: smf-prefix/src/smf-stamp/smf-configure] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:230: CMakeFiles/smf.dir/all] Error 2
gmake: *** [Makefile:84: all] Error 2
CMake Error at CMakeLists.txt:58 (message):
  Build step for v::build failed: 2

```

**Comment by senior7515** [comment]
@noahdesu anything comes to mind as to why w/ redpanda, the findc-ares.cmake isn't working. What would cause it ?

**Comment by noahdesu** [comment]
i hit this too, and it had me stumped for a while. the library is being found (via find_library), but pkg-config which can parse the version information isn't finding the .pc file. make sure that the search paths are setup correctly in your environment.

**Comment by nacl** [comment]
Something like this is definitely needed.  I would, however, suggest that the build management mode be defaulted to "off".

The invocation of `cmake` (or `configure`, etc) for a build is typically assumed to be a relatively fast, consequence-free configuration.  Having the configuration phase build all dependencies is rarely seen pattern; I was surprised to see it happen when I tried a build off master today.

The documentation should be updated soon in any case; it still references CMakeCooking.

**Comment by senior7515** [comment]
@noahdesu - what's the name of the cmake var for .pc paths? cmake module path is not it.

**Comment by senior7515** [comment]
@nacl yeah. good point on both accounts. 

can u file tix/issues for them. we should definitely fix documentation.

**Comment by noahdesu** [comment]
@nacl

> Having the configuration phase build all dependencies is rarely seen pattern

I 100% agree with this, and I can't really imagine anyone using this as-is in normal development. We wanted to move away from the cooking approach quickly, but also wanted an easy way to do isolated builds, so this worked out well.  But I think that what you're seeing right now with the build system can be considered step 1. Step 2 should definitely be to separate the build configurations so that one option is to maximize the amount of system-level dependencies being used.

**Comment by noahdesu** [comment]
@senior7515 `CMAKE_MODULE_PATH` is for locating things like `FindLibrary.cmake`. What you are after is `CMAKE_PREFIX_PATH` or the `PKG_CONFIG_PATH` environment variable.

Are you using `c-ares` directly and seeing this issue, or indirectly via `Seastar`'s dependency? If it is the later case, then it would be useful to know how you are approaching `smf` integration into your project, as there are several approaches, and at least a few that probably are untested. I think this is where the example project needs a little love.

**Comment by nacl** [comment]
@senior7515: Done: #315, #316 

@noahdesu That's fair.  I was just a bit surprised when I pulled a build from master and then a big pile of software builds came out of nowhere.  This PR (along with the above tickets) should come soon so that other developers are not surprised.

And indeed, CMake ExternalProject does a pretty good job of staying out of the way as long as it doesn't have to query for repository updates.  It would be cool to see how well the build does with the [Kitware Ninja branch](https://github.com/Kitware/ninja) that integrates with the Make Jobserver.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -21,13 +21,14 @@ install(TARGETS smfc
...
   DESTINATION ${CMAKE_INSTALL_BINDIR})
 
-# TODO: rpc_generated.h should not be checked into git. instead, it should be
-# generated at build-time and installed/packaged explicitly.
 smfc_gen(
   CPP
   TARGET_NAME raw_rpc_gen
   OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/include/smf
   SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/core/rpc.fbs)
+install(
```
Comment (senior7515) [inline src/CMakeLists.txt]: @noahdesu  this was the func missing on issue #317

Comment (noahdesu) [reply src/CMakeLists.txt]: Yep. Looks like you are tackling some/all of the installation stuff I didn't get around to in the cmake-ng PR. I'll build tickets for the remainder of the todo's there.

Comment (senior7515) [reply src/CMakeLists.txt]: yeah, i just wanted to fix the install() issues at least for redpanda on a speedy schedule. thanks for following up tho ! :) :+1:  :100:

**Pull request merged**